gangplank/ocp/cosa-pod.go

```diff
@@ -89,48 +96,89 @@ var (
 	}
 )
 
-// setPodDefaults checks the Kubernetes version to determine if
-// we're on OCP 3.11 (v1.11) or 4.2(v1.15) or later.
-// https://docs.openshift.com/container-platform/4.2/release_notes/ocp-4-2-release-notes.html#ocp-4-2-about-this-release
-func setPodDefaults() error {
-	vi, err := apiClientSet.DiscoveryClient.ServerVersion()
-	if err != nil {
-		return fmt.Errorf("failed to query the kubernetes version: %w", err)
-	}
+// cosaPod is a COSA pod
+type cosaPod struct {
+	apiBuild     *buildapiv1.Build
+	apiClientSet *kubernetes.Clientset
+	project      string
 
-	minor, err := strconv.Atoi(strings.TrimRight(vi.Minor, "+"))
-	log.Infof("Kubernetes version of cluster is %s %s.%d", vi.String(), vi.Major, minor)
-	if err != nil {
-		return fmt.Errorf("failed to detect OCP cluster version: %v", err)
+	ocpInitCommand  []string
+	ocpRequirements v1.ResourceList
+	ocpSecContext   *v1.SecurityContext
+	volumes         []v1.Volume
+	volumeMounts    []v1.VolumeMount
+
+	index int
+	pod   *v1.Pod
+}
+
+// CosaPodder create COSA capable pods.
+type CosaPodder interface {
+	WorkerRunner(ctx context.Context, envVar []v1.EnvVar) error
+	PodmanRunner(ctx context.Context, workDir string, envVar []v1.EnvVar) error
+}
+
+// a cosaPod is a CosaPodder
+var _ = CosaPodder(&cosaPod{})
+
+// NewCosaPodder creates a CosaPodder
+func NewCosaPodder(
+	ctx context.Context,
+	apiBuild *buildapiv1.Build,
+	apiClientSet *kubernetes.Clientset,
+	project string,
+	index int) (CosaPodder, error) {
+
+	cp := &cosaPod{
+		apiBuild:     apiBuild,
+		apiClientSet: apiClientSet,
+		project:      project,
+		index:        index,
+
+		// Set defaults for OCP4.x
+		ocpRequirements: ocpRequirements,
+		ocpSecContext:   ocpSecContext,
+		ocpInitCommand:  ocpInitCommand,
+
+		volumes:      volumes,
+		volumeMounts: volumeMounts,
 	}
-	if minor >= 15 {
-		log.Info("Detected OpenShift 4.x cluster")
-		return nil
+
+	if cp.apiClientSet != nil {
+		vi, err := cp.apiClientSet.DiscoveryClient.ServerVersion()
+		if err != nil {
+			return nil, fmt.Errorf("failed to query the kubernetes version: %w", err)
+		}
+
+		minor, err := strconv.Atoi(strings.TrimRight(vi.Minor, "+"))
+		log.Infof("Kubernetes version of cluster is %s %s.%d", vi.String(), vi.Major, minor)
+		if err != nil {
+			return nil, fmt.Errorf("failed to detect OCP cluster version: %v", err)
+		}
+		if minor >= 15 {
+			log.Info("Detected OpenShift 4.x cluster")
+			return cp, nil
+		}
+
+		log.Infof("Creating container with Openshift v3.x defaults")
```

Comment:
(Minor but in this short span of code there is all 4 of "kubernetes", "OCP", "OpenShift", "Openshift"; I'd probably stick with "Kubernetes" and "OpenShift")

Reply:
fixed up